### DISCUSSION TITLE
Migrating to `nodemailer` to update Nest to version 10

### DISCUSSION THREAD

**Discussion by ahmadalfy** [nest-modules/mailer#998] [General]
Hello,

If you are stuck with `@nest-modules/mailer` and need to update Nest to version 10, migrating to `nodemailer` is very simple and doesn't require much change to your code.

First you need to create a mailer service and replace `@nest-modules/mailer` MailerService with it. It's straight forward.

Create a file and name it `mailer.service.ts`

```typescript
import { Injectable } from '@nestjs/common';
import * as nodemailer from 'nodemailer';

@Injectable()
export class MailerService {

	private transporter: nodemailer.Transporter

	constructor() {
		this.transporter = nodemailer.createTransport({
			host: process.env.MAILER_HOST,
			port: Number(process.env.MAILER_PORT),
			secure: process.env.MAILER_SECURE === 'true',
			auth: {
				user: process.env.MAILER_USER,
				pass: process.env.MAILER_PASSWORD,
			},
		}, {
			from: process.env.MAILER_FROM,
		});
	}

	async sendMail(mailOptions: nodemailer.SendMailOptions) {
		return this.transporter.sendMail(mailOptions);
	}
}
```

You need to import this service into your `app.module.ts` then you can easily inject that service into anything in your app.

Remove `@nest-modules/mailer` and its references in the app module and the rest of your application.

Once an update is released, you can install it back and use it if you want.

**Comment by mkgn** [comment]
Using this I got the basic mail sending to work properly. How can I configure the template directories for handlebars if i am going to use it? nestjs/mailer seems to support that and I know your suggestion is a fix until the versions are updated. As of now it seems I will have to use a nodemailer different package to support templating? How did you manage the templating features?

**Reply by ahmadalfy** [reply]
I wasn't using the templating feature as the emails I was sending were simple. I didn't dig into that part I am sorry.

**Reply by ahmadalfy** [reply]
@mkgn check [this comment](https://github.com/nest-modules/mailer/issues/997#issuecomment-1611815041), it allows you to override the resolution and it will eventually work

**Reply by ecorona** [reply]
You could write a few extra lines to parse your own template (ejs, hbs, etc) in to html and then pass that to mail options.

With hbs:

```TS
import { compile } from 'handlebars';
//...
//read and parse hbs template
const templateContent = readFileSync(__dirname + '/templates/template.hbs').toString();
const hbTemplate = compile(templateContent);
mailOptions.html = hbTemplate({ foo: 'bar' });
//... send email
this.transporter.sendMail(mailOptions);
```
With ejs:

```TS
import * as ejs from 'ejs';
//...
//parse ejs template
mailOptions.html = await ejs.renderFile(__dirname + '/templates/template.ejs', { foo: 'bar' });
//... send email
this.transporter.sendMail(mailOptions);
```

**Comment by juandav** [comment]
At this moment, the mailer is already compatible with version 10 of Nest.
